pkg/controller/v1alpha1/localmodelnode/controller.go

```diff
@@ -45,13 +55,237 @@ type LocalModelNodeReconciler struct {
+		}
+	}
+
+	if err := c.Status().Update(context.Background(), localModelNode); err != nil {
+		c.Log.Error(err, "Update local model cache status error", "name", localModelNode.Name)
+		return err
+	}
+
+	return nil
+}
+
 func (c *LocalModelNodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	c.Log.Info("Daemonset controller reconciling localmodelnode", "name", req.Name)
+	c.Log.Info("Agent reconciling LocalModelNode", "name", req.Name, "node", nodeName)
+	// Create Jobs to download models if the model is not present locally.
+	// 1. Check if LocalModelNode CR is for current node
+	localModelNode := &v1alpha1api.LocalModelNode{}
+	if err := c.Get(ctx, req.NamespacedName, localModelNode); err != nil {
+		c.Log.Error(err, "Error getting LocalModelNode", "name", req.Name)
+		return reconcile.Result{}, err
+	}
+	if nodeName != localModelNode.ObjectMeta.Name {
+		c.Log.Info("Skipping LocalModelNode because it is not for current node", "name", req.Name, "current node", nodeName, "intended node", localModelNode.ObjectMeta.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// 2. Kick off download jobs for all models in spec
+	localModelConfig, err := v1beta1.NewLocalModelConfig(c.Clientset)
+	if err != nil {
+		c.Log.Error(err, "Failed to get local model config")
+		return reconcile.Result{}, err
+	}
+	defaultJobImage = localModelConfig.DefaultJobImage
+	FSGroup = localModelConfig.FSGroup
+	if err := c.DownloadModel(ctx, localModelNode, localModelConfig.JobNamespace); err != nil {
+		c.Log.Error(err, "Model download err")
+	}
+	// Delete models from disk if they are not on the CR.
+	// 1. Scan model dir and get a list of downloaded models
+	entries, err := os.ReadDir(localModelContainerVol + "/models")
+	if err != nil {
+		c.Log.Error(err, "Failed to list model directory", "dir", localModelContainerVol+"/models")
+	}
+	localModels := map[string]struct{}{}
+	for _, entry := range entries {
+		// Models could only exist in sub dir
+		if entry.IsDir() {
+			localModels[entry.Name()] = struct{}{}
+		}
+	}
+	c.Log.Info("Found existing local models", "models", maps.Keys(localModels))
+
+	// 2. Compare with list of models from LocalModelNode CR
+	for _, localModelInfo := range localModelNode.Spec.LocalModels {
+		// Remove expected models from local model set
+		delete(localModels, localModelInfo.ModelName)
+	}
+	// 3. Models not in LocalModelNode CR spec should be deleted
+	if len(localModels) != 0 {
+		c.Log.Info("Removing models", "models", maps.Keys(localModels))
+		for localModelName := range localModels {
+			modelDir := localModelContainerVol + "/models/" + localModelName
+			if err := os.RemoveAll(modelDir); err != nil {
+				c.Log.Error(err, "Failed to remove model directory", "dir", modelDir)
+			}
+		}
+	}
+
 	return reconcile.Result{}, nil
 }
 
 func (c *LocalModelNodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1api.LocalModelNode{}).
```

Comment:
(different pr) might be better filter by node name in the predicate  https://github.com/kubernetes-sigs/controller-runtime/blob/c1331a5c7c5880c35223d5dec0d87ac166554d46/pkg/builder/controller.go#L83